plugins/modules/artisan.py

```diff
@@ -0,0 +1,324 @@
+    command: "{{ item }}"
+    options:
+      - database=sqlite
+      - isolated=true
+  loop:
+    - migrate
+    - seed
+
+# php artisan up --no-ansi --no-interaction
+- name: Clear caches and bring the application out of maintenance mode
+  community.general.artisan:
+    working_dir: "{{ project_src }}"
+    command: "{{ item }}"
+  loop:
+    - cache:clear
+    - event:clear
+    - route:clear
+    - view:clear
+    - up
+'''
+
+import os
+import re
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+def parse(out):
+    # remove any unicode strings
+    pattern = re.compile(r'(?:\x1B[@-_]|[\x80-\x9F])[0-?]*[ -/]*[@-~]')
+
+    # split and trim lines
+    lines = list(filter(None, out.splitlines()))
+    message = stderr = lines[0]
+
+    # second line has the exception message
+    if len(lines) > 1:
+        stderr = lines[1]
+
+    # message, stderr, stderr_lines
+    return [
+        re.sub(pattern, '', message).strip(),
+        re.sub(pattern, '', stderr).strip(),
+        list(map(str.strip, lines))
+    ]
+
+
+def has_changed(message):
+    # message fragments that Laravel gives us
+    artisan_messages = [
+        'No publishable',  # resources for tag
+        'Nothing to',  # migrate,...
+        'already',  # up, down, exists,...
+        'Not modified',
+        'No changes'
+    ]
+
+    for info in artisan_messages:
+        if info in message:
+            return False
+
+    return True
+
+
+def get_help(module, command):
+    # get all available options and arguments from an artisan command using artisan help to json
+    rc, out, err = artisan_command(module, 'help %s' % command, options='--format=json --no-ansi')
+
+    # no command, no help, no json
+    if rc != 0:
+        module.fail_json(msg='%s is not part of the command definitions.' % command)
+
+    return module.from_json(out)
+
+
+def get_options(module, command_help):
+    # set available command options
+    available_options = command_help['definition']['options']
+
+    # merge default options
+    command_options = module.params['options'] + [
+        'no-interaction',
+        'no-ansi',
+        'force',
+    ]
+
+    options = []
+
+    # check availability and format options
+    for option in command_options:
+        option = list(map(str.strip, option.lstrip('-').split('=')))
+        if option[0] in available_options:
+            options.append('--%s' % '='.join(option))
+
+    # return the unique options turnkey ready as a string
+    return ' '.join(list(set(options))).strip()
+
+
+def get_arguments(module, command_help):
+    # set available command arguments
+    available_arguments = command_help['definition']['arguments']
+
+    # mostly we have no arguments
+    arguments = ''
+
+    # if arguments expected for command, join from params
+    if len(available_arguments) > 0:
+        arguments = ' '.join(module.params['args'])
+
+    # return arguments turnkey ready as a string
+    return arguments.strip()
```

Comment:
I failed to see the usefulness of this. By doing it - and I am assuming quite a bit from the code, I am not familiar with `artisan` - my understanding is that the module will only return the concatenated `args` if the command expects arguments. Otherwise, those `args` are silently ignored.

Wouldn't it be better if we simply pass the `args` as is to user, and if they did something wrong they get an error message? The code as it is prevents users from knowing that they passed invalid parameters, and the module seems a bit over-engineered to make that happen. `get_options` and `get_help` could possibly disappear as well.

Reply:
That's exactly how it was intended. The module itself uses a default argument, which I'm not sure is always present. In that case an exception would be thrown. However, if we went with our own Ansible modules and state, as described below, the args would no longer be parsed.